Dpmma 1743 focus actions on contact timeline

AlanWierzchonCA opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ no]
| New feature/enhancement? (use the a.x branch)      | [ yes]
| Deprecations?                          | [ no]
| BC breaks? (use the c.x branch)        | [no ]
| Automated tests included?              | [ yes] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--






Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

This PR adds Focus Item events like view, click in Contacts timeliene.

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create Focus Item with trackable link
3. Identify contact 
4. Go to page with working Focus Item
5. Check contact timeline
6. Click link from Focus Item
7. Check contact timeline

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->
![image](https://user-images.githubusercontent.com/100592709/220212174-a30bd9b9-5a3b-47f7-bc16-d81f6cb9f344.png)

AlanWierzchonCA committed:
feat: [DPMMA-1743] add focus view to contact timeline

AlanWierzchonCA committed:
feat: [DPMMA-1743] add focus click to contact timeline cdn

AlanWierzchonCA committed:
fix: [DPMMA-1743] fix filters on contact timeline

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix twig contact timeline

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix deprecated translator interface

AlanWierzchonCA committed:
fix:[DPMMA-1743] functional test for LeadSubscriber.php in MauticFocusBundle

AlanWierzchonCA committed:
fix:[DPMMA-1743] functional test click FI for LeadSubscriber.php in MauticFocusBundle

AlanWierzchonCA committed:
feat:[DPMMA-1743] search focus events

AlanWierzchonCA committed:
fix:[DPMMA-1743] fixes phpstan and cs-fixer

AlanWierzchonCA committed:
refactor:[DPMMA-1743] remove else with continue

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix phpstan error

kuzmany reviewed (dismissed):
Looks good to me 👍

escopecz commented on plugins/MauticFocusBundle/EventListener/LeadSubscriber.php:
```diff
@@ -0,0 +1,108 @@
+    /*
+     * @var FocusModel
+     */
+    private FocusModel $focusModel;
```
```suggestion
    private Translator $translator;

    private RouterInterface $router;

    private FocusModel $focusModel;
```

AlanWierzchonCA replied:
I changed this.

escopecz commented on plugins/MauticFocusBundle/Entity/StatRepository.php:
```diff
@@ -39,4 +39,37 @@ public function getStats($id, $type, $fromDate = null)
+
+        $result = $q->getQuery()->getArrayResult();
+
+        return ['result' => $result, 'total' => count($result)];
```
Why does it return the count? The method that calls this method can do the count on the array too, right?

Also, I'm a bit afraid doing select queries without any limit. Check other timeline implementations. They use `trait TimelineTrait::getTimelineResults()` to build the filtering.

AlanWierzchonCA replied:
Hi @escopecz,
paging these events did not work. I fixed this.

Best,
Alan

AlanWierzchonCA committed:
fix:[DPMMA-1743] delete annotations types in Lead Subscriber class

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix pagination focus events on contact timeline

AlanWierzchonCA committed:
resolve conflict 5.x

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix tests

AlanWierzchonCA committed:
Merge branch '5.x' into DPMMA-1743_FIaction_on_contactHistory_5x

escopecz commented on plugins/MauticFocusBundle/EventListener/LeadSubscriber.php:
```diff
@@ -0,0 +1,96 @@
+        $statsClickByLead = $this->focusModel->getStatRepository()->getStatsClickByLead($contactId, $event->getQueryOptions());
+
+        if (!$event->isEngagementCount()) {
+            $template = 'MauticFocusBundle:SubscribedEvents\Timeline:index.html.php';
```
The PHP engine was removed from the 5.x branch. Please use html.twig files only. But in this case I wonder if we can omit the template completely when it's empty.

AlanWierzchonCA replied:
I deleted this template.

escopecz commented on plugins/MauticFocusBundle/Entity/StatRepository.php:
```diff
@@ -39,4 +42,53 @@ public function getStats($id, $type, $fromDate = null)
+
+        if (isset($options['search']) && $options['search']) {
+            $q->andWhere($q->expr()->orX(
+                $q->expr()->like('f.name', $q->expr()->literal('%'.$options['search'].'%')),
+                $q->expr()->like('f.description', $q->expr()->literal('%'.$options['search'].'%')),
+                $q->expr()->like('s.type', $q->expr()->literal('%'.$options['search'].'%'))
```
Be careful with queries like this as searching with wildcards from the beginning is expensive as it cannot use indexes. So if the table grows into hundreds of thousands then searching will become unusable as it will time out the request and cause issues for the users. In many places we use only wildcard at the end of the string.

Also, does it make sense to search in the type? It's already handled above. But why the "like" operator was used there?

AlanWierzchonCA replied:
Hi @escopecz, 
In the timeline search, page hits are also searched in this way. 
```
        // app/bundles/PageBundle/Entity/HitRepository.php  
        // Line: 86
        $query->andWhere($query->expr()->like('p.title', $query->expr()->literal('%'.$options['search'].'%')));        
```
e.g. 

without a wildcard from the beginning, there will be no focus named "[DPMMA-1743] focus test" in the result:
![image](https://user-images.githubusercontent.com/100592709/229822137-928d84ea-4866-4a2f-a202-a51891fa20ca.png)

ant with a wildcard from the beginning yes: 
![image](https://user-images.githubusercontent.com/100592709/229823162-5fb19f38-0c58-49f0-bd4f-731c97cab215.png)

A type search makes sense when its search phrase "click" or "view":
![image](https://user-images.githubusercontent.com/100592709/229817344-e3c8a66e-299d-4f71-9690-8a4775792c14.png)

but It doesn't sense with "like" operator. I Changed this. 

In conclusion you are right that searching with wildcard is expensive. However, removing the wildcard from the beginning will limit the search possibilities. I leave the final decision in this matter to you :).

AlanWierzchonCA committed:
fix:[DPMMA-1743]  remove focus timeline template

AlanWierzchonCA committed:
fix:[DPMMA-1743]  remove focus timeline template: fix unitTests

AlanWierzchonCA committed:
fix:[DPMMA-1743] delete serach type wildcards

AlanWierzchonCA committed:
feat:[DPMMA-1743] functional tests for searching events on contact timeline

AlanWierzchonCA commented:
Hi @escopecz, 
I added functional tests.

Best, 
Alan

escopecz commented on plugins/MauticFocusBundle/Entity/Stat.php:
```diff
@@ -38,6 +38,9 @@ class Stat
     private $dateAdded;
 
+    /**
+     * @var Lead
```
```suggestion
     * @var ?Lead
```
When you create a fresh entity then this is null. 
```php
$stat = new Stat();
$stat->getLead(); // returns null, not Lead
```

AlanWierzchonCA replied:
I changed this.

escopecz commented on plugins/MauticFocusBundle/Entity/StatRepository.php:
```diff
@@ -39,4 +42,53 @@ public function getStats($id, $type, $fromDate = null)
+
+        $q->where($q->expr()->andX(
+            $q->expr()->eq('s.lead_id', (int) $leadId),
+            $q->expr()->like('s.type', '"'.$type.'"')
```
I thought you wrote that you've changed the `like` into `eq` but it's still there. [I've read](https://stackoverflow.com/questions/543580/equals-vs-like) that `like` is 300 times slower compared to `eq`. It will differ based on the amount of data in the table, but still. Why not make it faster?
```suggestion
            $q->expr()->eq('s.type', ':type');
```
And somewhere bellow you'll set the param:
```php
$q->setParameter('type', $type);
```

AlanWierzchonCA replied:
I changed this.

escopecz commented on plugins/MauticFocusBundle/Entity/StatRepository.php:
```diff
@@ -39,4 +42,53 @@ public function getStats($id, $type, $fromDate = null)
+            $q->andWhere($q->expr()->orX(
+                $q->expr()->like('f.name', $q->expr()->literal('%'.$options['search'].'%')),
+                $q->expr()->like('f.description', $q->expr()->literal('%'.$options['search'].'%')),
+                $q->expr()->like('s.type', $q->expr()->literal($options['search']))
```
Same here. Why use `like` over `eq`?

AlanWierzchonCA replied:
I changed it for search in 's.type'. I leave 'like' for wildcard searches.

escopecz commented on plugins/MauticFocusBundle/Entity/StatRepository.php:
```diff
@@ -39,4 +42,53 @@ public function getStats($id, $type, $fromDate = null)
+
+        if (isset($options['search']) && $options['search']) {
+            $q->andWhere($q->expr()->orX(
+                $q->expr()->like('f.name', $q->expr()->literal('%'.$options['search'].'%')),
+                $q->expr()->like('f.description', $q->expr()->literal('%'.$options['search'].'%')),
```
I'm afraid I can't let this be merged. We've done this mistake many times and it resulted in angry customers that cannot load a page when they start searching. If I wouldn't learn from past mistakes then what good am I?

To make a compromise between usablity and performance, change this to
```suggestion
                $q->expr()->like('f.name', $q->expr()->literal($options['search'].'%')),
                $q->expr()->like('f.description', $q->expr()->literal($options['search'].'%')),
```
And add indexes on the `Stat` entity on these 2 columns + migration for existing Mautic instances adding those indexes.

AlanWierzchonCA replied:
No problemo :) 
I changed the wildcard for 'f.name' and removed the 'f.description' search. 
I added the index and migration 'f.name'

escopecz reviewed (changes requested):
The tests look great! It might be less work and making sure that it all works together if you'd load the contact detail page and read the timeline content instead of dispatching the event, but this is good too.

AlanWierzchonCA committed:
feat:[DPMMA-1743] change search wildcard

AlanWierzchonCA committed:
feat:[DPMMA-1743] adding focus index migrations and optimizing the search query

escopecz reviewed (dismissed):
Thank you! I see no other issues with this code 👍

Syrgak-Alan commented:
@AlanWierzchonCA  hi, I tried to test your pull request several times, but for some reason, clicks on the focus item are not counted for me. Contact actions look like this:
~ Page hit
~ Focus view
~ Page hit(but here expected focus clicked)
what could be a problem? 
<img width="1280" alt="Bildschirmfoto 2023-06-01 um 16 43 48" src="https://github.com/mautic/mautic/assets/96595532/6f9c42fc-a49c-41ef-9772-6d27a05cf310">

shinde-rahul reviewed (approved):
LGTM!

RCheesley reviewed (changes requested):
I dropped this into a landing page and tested in an incognito window, but I am seeing double clicks being reported (I deffo only clicked once):

![screenshot-8080-alanwierzchonca-mautic-lwfr2zt03pf ws-eu99 gitpod io-2023 06 09-15_17_53](https://github.com/mautic/mautic/assets/2930593/65379e55-f434-41d0-adcb-e677df468879)

Also despite the fact that I am reloading the page every time, I only see one instance of the focus item being viewed - is this maybe just unique views, or should it fire a 'view' each time the page is loaded?

I'm using a bar focus item and it's showing each time I load.

AlanWierzchonCA commented:
I will check it!

RCheesley commented:
Some conflicts to be resolved now @AlanWierzchonCA if you can check them while addressing the feedback please! Thank you!

AlanWierzchonCA commented:
I'll deal with that next week

AlanWierzchonCA committed:
feat:[DPMMA-1743] merge with 5.x changes

AlanWierzchonCA committed:
Merge branch '5.x' into DPMMA-1743_FIaction_on_contactHistory_5x_merge

AlanWierzchonCA committed:
feat:[DPMMA-1743] fix deprecatd method orX endX

AlanWierzchonCA commented:
Hi, @RCheesley,

I was unable to reproduce the case of double clicks being reported.
I updated changes from version 5.x, fixed conflicts and cs fixers & phpstan errors.

![fi_contact_stats](https://github.com/mautic/mautic/assets/100592709/45ccbe31-00e2-40d7-9648-662ecd121077)

![fi_bar_config](https://github.com/mautic/mautic/assets/100592709/52eaac2e-165c-44f7-8f6d-3cfb2abb8a9c)

AlanWierzchonCA reviewed (commented):
LGTM!

AlanWierzchonCA committed:
Update config.yaml - undo changes

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix cs fixers errors

AlanWierzchonCA committed:
fix:[DPMMA-1743] fix cs fixers errors

RCheesley reviewed (approved):
This seems to be working now no longer seeing the double entries on the timeline as I was before:

![screenshot-8080-alanwierzchonca-mautic-4s78x9mjwto ws-eu105 gitpod io-2023 10 03-14_47_00](https://github.com/mautic/mautic/assets/2930593/3f68b800-c49c-44e1-8bef-89dbd9615153)

Good to go! 🚀

AlanWierzchonCA commented:
I'm glad :D

escopecz commented:
@AlanWierzchonCA this broke the API tests. See
- https://github.com/mautic/api-library/actions/runs/6404467617/job/17385557164
- https://github.com/mautic/api-library/blob/b3258bcf1a87ff45bbaf6b17fe5ce7e50b1f5843/tests/Api/ContactsTest.php#L189

Error message: 
```
[2023-10-04 10:18:52] mautic.CRITICAL: Uncaught PHP Exception Error: "Call to a member function getId() on null" at /var/www/html/plugins/MauticFocusBundle/EventListener/LeadSubscriber.php line 57 {"exception":"[object] (Error(code: 0): Call to a member function getId() on null at /var/www/html/plugins/MauticFocusBundle/EventListener/LeadSubscriber.php:57)
[stacktrace]
#0 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(270): MauticPlugin\\MauticFocusBundle\\EventListener\\LeadSubscriber->onTimelineGenerate()
#1 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(230): Symfony\\Component\\EventDispatcher\\EventDispatcher::Symfony\\Component\\EventDispatcher\\{closure}()
#2 /var/www/html/vendor/symfony/event-dispatcher/EventDispatcher.php(59): Symfony\\Component\\EventDispatcher\\EventDispatcher->callListeners()
#3 /var/www/html/app/bundles/LeadBundle/Model/LeadModel.php(2067): Symfony\\Component\\EventDispatcher\\EventDispatcher->dispatch()
#4 /var/www/html/app/bundles/LeadBundle/Controller/Api/LeadApiController.php(396): Mautic\\LeadBundle\\Model\\LeadModel->getEngagements()
#5 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(163): Mautic\\LeadBundle\\Controller\\Api\\LeadApiController->getAllActivityAction()
#6 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(75): Symfony\\Component\\HttpKernel\\HttpKernel->handleRaw()
#7 /var/www/html/vendor/symfony/http-kernel/Kernel.php(202): Symfony\\Component\\HttpKernel\\HttpKernel->handle()
#8 /var/www/html/app/AppKernel.php(104): Symfony\\Component\\HttpKernel\\Kernel->handle()
#9 /var/www/html/app/middlewares/CORSMiddleware.php(82): AppKernel->handle()
#10 /var/www/html/app/middlewares/HSTSMiddleware.php(35): Mautic\\Middleware\\CORSMiddleware->handle()
#11 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(34): Mautic\\Middleware\\HSTSMiddleware->handle()
#12 /var/www/html/app/middlewares/VersionCheckMiddleware.php(58): Mautic\\Middleware\\CatchExceptionMiddleware->handle()
#13 /var/www/html/app/middlewares/TrustMiddleware.php(42): Mautic\\Middleware\\VersionCheckMiddleware->handle()
#14 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\TrustMiddleware->handle()
#15 /var/www/html/index.php(19): Stack\\StackedHttpKernel->handle()
#16 {main}
```
Will you have time to look into it?

escopecz commented:
I'm going to revert this PR tomorrow if there won't be no one to fix the regression.

AlanWierzchonCA commented:
I will take it.

AlanWierzchonCA commented:
@escopecz, 
I'm going to create a new PR for fixing the api library case.
ok?

escopecz commented:
That'd be amazing!

AlanWierzchonCA commented:
I deal with this in PR:
https://github.com/mautic/mautic/pull/12774
